datetime/datetime.go

```diff
@@ -0,0 +1,137 @@
+// Package with support of Tarantool's datetime data type.
+//
+// Datetime data type supported in Tarantool since 2.10.
+//
+// See also:
+//
+// * Datetime Internals https://github.com/tarantool/tarantool/wiki/Datetime-Internals
+//
+// Since: 1.6
+package datetime
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+	"time"
+
+	"encoding/binary"
+
+	"gopkg.in/vmihailenco/msgpack.v2"
+)
+
+// Datetime MessagePack serialization schema is an MP_EXT extension, which
+// creates container of 8 or 16 bytes long payload.
+//
+//   +---------+--------+===============+-------------------------------+
+//   |0xd7/0xd8|type (4)| seconds (8b)  | nsec; tzoffset; tzindex; (8b) |
+//   +---------+--------+===============+-------------------------------+
+//
+// MessagePack data encoded using fixext8 (0xd7) or fixext16 (0xd8), and may
+// contain:
+//
+// * [required] seconds parts as full, unencoded, signed 64-bit integer,
+// stored in little-endian order;
+//
+// * [optional] all the other fields (nsec, tzoffset, tzindex) if any of them
+// were having not 0 value. They are packed naturally in little-endian order;
+
+// Datetime external type
+// Supported since Tarantool 2.10. See more details in issue
+// https://github.com/tarantool/tarantool/issues/5946
+const datetime_extId = 4
+
+/**
+ * datetime structure keeps number of seconds and
+ * nanoseconds since Unix Epoch.
+ * Time is normalized by UTC, so time-zone offset
+ * is informative only.
+ */
+type datetime struct {
+	// Seconds since Epoch
+	seconds int64
+	// Nanoseconds, fractional part of seconds
+	nsec int32
+	// Timezone offset in minutes from UTC
+	// (not implemented in Tarantool, see gh-163)
+	tzOffset int16
+	// Olson timezone id
+	// (not implemented in Tarantool, see gh-163)
+	tzIndex int16
+}
+
+const (
+	secondsSize  = 8
+	nsecSize     = 4
+	tzIndexSize  = 2
+	tzOffsetSize = 2
+)
+
+func encodeDatetime(e *msgpack.Encoder, v reflect.Value) error {
+	var dt datetime
+
+	tm := v.Interface().(time.Time)
+	dt.seconds = tm.Unix()
+	dt.nsec = int32(tm.Nanosecond())
+	dt.tzIndex = 0  /* not implemented, see gh-163 */
+	dt.tzOffset = 0 /* not implemented, see gh-163 */
+
+	var bytesSize = secondsSize
+	if dt.nsec != 0 || dt.tzOffset != 0 || dt.tzIndex != 0 {
+		bytesSize += nsecSize + tzIndexSize + tzOffsetSize
+	}
+
+	buf := make([]byte, bytesSize)
+	binary.LittleEndian.PutUint64(buf[0:], uint64(dt.seconds))
+	if bytesSize == 16 {
+		binary.LittleEndian.PutUint32(buf[secondsSize:], uint32(dt.nsec))
+		binary.LittleEndian.PutUint16(buf[secondsSize+nsecSize:], uint16(dt.tzOffset))
+		binary.LittleEndian.PutUint16(buf[secondsSize+nsecSize+tzOffsetSize:], uint16(dt.tzIndex))
```

Comment:
Are these signed or unsigned values? Why do we need to convert int -> uint when encoding and uint -> int when decoding? I don't see such code in the C implementation, just memcpy.

In your comment above:
```
[required] seconds parts as full, unencoded, signed 64-bit integer, stored in little-endian order;
```

Reply:
In Tarantool seconds defined as a `double` (signed) and nanoseconds as `int32` (unsigned), see https://github.com/tarantool/tarantool/blob/master/src/lib/core/datetime.h#L56-L71. Tarantool has a reason to prefer `double` instead of `in64`, see https://github.com/tarantool/tarantool/wiki/Datetime-internals#intervals-in-c:

> You might ask: why there are `int32_t` used by some fields, but `double`s by others?
> We switch to `double` once we need to support range of values wider than +-2billions.
> We would be happily use `int64_t` here instead of `double`, but `int64_t` is boxed 
> in LuaJIT ffi, thus would be increasing GC pressure. So after some benchmarking
> https://gist.github.com/tsafin/618fbf847d258f6e7f5a75fdf9ea945b it was decided that 
> `double` would be fast enough, especially if there AVX enabled in compiler.

So we can use `int64` for seconds.

For reading values from buffer, I use a package `binary` that has only unsigned variants of methods, see https://pkg.go.dev/encoding/binary#ByteOrder. Seems I need to restore a sign somehow after reading a value or use plain `Read()` and `Write()` and explicitly specify data types.

Reply:
It seems to me that using binary.Read/binary.Write will be less confusing + There will be no need to manually handle offsets and types.

Encode/decode tests with values close to INT_MAX/INT_MIN (for seconds) /INT32_MAX/INT32_MIN (for nanoseconds) are required to make sure that the operations are correct.

Reply:
> Encode/decode tests with values close to INT_MAX/INT_MIN (for seconds) /INT32_MAX/INT32_MIN (for nanoseconds) are required to make sure that the operations are correct. (Despite the fact that such values are logically impossible now)

Sure. Already added (skipped for now).

Reply:
Tests fixed.